### ISSUE TITLE
Gamepad left stick left & right wrong

### ISSUE THREAD

**Issue by EGGamer** [Unity-Technologies/InputSystem#356]
I have a composite DPad. Up is **leftstick/up**, Down is **leftstick/down**, Left...., Right.... Then I put that as a Vector2. What happens is that up is 0,1; and right is 1,0. Left should be -1,0 and down 0,-1, but left and down are both 0,0, so thats incorrect.

**Comment by XZcutable** [comment]
I am having the same issue.

If it helps I'm using an Xbox-One controller

**Comment by XZcutable** [comment]
lol well, that's why. Ok, you guys, in the Axis Composite class in the UnityEngine.Experimental.Input.Composites namespace you set negativeIsPressed equal to if the negativeValue is greater than 0. I just check and the value is between 0 and -1 so that will always evaluate to false. 

A similar issue is when you check to see which side wins, you return the -negativeValue which would actually make it positive (assuming how the other ones work, I think this is unintended).  You have the exact same issue when you return if negativeIsPressed as you negate the negativeValue, thus making it positive. I assume its the same issue in the Dpad Composite.

Edit:
~~For the Dpad, it seems to be slightly more complicated. I sussed out that when the left stick is pressed whatever calls the MakeDpadVector sets left to false. Still trying to figure out which class is calling this.~~
Actually it seems that the reason left is false is because when ReadValue(ref InputBindingCompositeContext context) in DpadComposite is called, leftValue when pressed.

Edit:
Ok I think I understand why the issue arrised in the first place. Anlalog passes a value between -1 and 1 while the keyboard passes a value between 0 or 1 (or is it 0 or 1). So a way to distinguish the two is needed.

**Comment by EGGamer** [comment]
No. I'm debugging the values into de console, and keyboard does pass -1,0;0,-1. But the stick does not for some reason. I'm also using an XBOX controller BTW.

**Comment by XZcutable** [comment]
The keyboard does pass -1 0 and 1 when you use the Axis Composite class, but the logic for the keyboard does not work for the gamepad analog stick. I am 100% certain as I debugged values by editing the scripts and made a fix for the Axis Composite for the Analog-stick but it breaks the keyboard input.

**Comment by EGGamer** [comment]
In theory, the analog stick should pass -1,0, 0,-1:
![2018-12-29_16-12-33](https://user-images.githubusercontent.com/40073654/50539690-98e14700-0b84-11e9-8a53-4abfc468ba7e.png)
![2018-12-29_16-12-49](https://user-images.githubusercontent.com/40073654/50539691-98e14700-0b84-11e9-8ad2-d10723809287.png)

EDIT: **It puts Xbox 360 controller, but it's an Xbox one controller**

**Comment by EGGamer** [comment]
@Rene-Damm

**Comment by Rene-Damm** [comment]
Sorry for the very long delay here.

>What happens is that up is 0,1; and right is 1,0. Left should be -1,0 and down 0,-1, but left and down are both 0,0, so thats incorrect.

This is as intended. The `up`, `down`, `right`, and `left` controls act as buttons and behave like their corresponding controls in `dpad`. Each button goes from 0 to 1 and represents a half-axis.

To make individual buttons go [0..-1] simply put an `Invert` processor on the binding.

>In theory, the analog stick should pass -1,0, 0,-1:

Just to confirm, when using an Xbox controller, you see the `gamepad.leftStick` and `gamepad.rightStick` controls not go [-1..1] on both X and Y? Is this with WebGL?

**Comment by EGGamer** [comment]
> Just to confirm, when using an Xbox controller, you see the `gamepad.leftStick` and `gamepad.rightStick` controls not go [-1..1] on both X and Y? Is this with WebGL?

When the leftstick is at the left and down it returns **0, 0**, that should't be happening. And that's a website that I've found on google.

**Comment by danielheim** [comment]
I've been experiencing a similar issue, using a standard action binding, bound to **leftStick/up** and **leftStick/down**

> Just to confirm, when using an Xbox controller, you see the gamepad.leftStick and gamepad.rightStick controls not go [-1..1] on both X and Y? Is this with WebGL?

I've written some crude test code
```
void InitGamepadTest()
{
	var action = new UnityEngine.Experimental.Input.InputAction(binding: "/<gamepad>/leftStick/*");

	action.performed += ctx =>
	{
		var val = ctx.ReadValue<float>();
		// filter here
		if (ctx.control.name == "up")
			Debug.Log($"{ctx.control}; value: {val}");
	};

	action.Enable();
}
```

Filtering by `"x"` and `"y"`, values from a DualShockGampadHID (Wired DualShock4) come through from [-1...1] as expected. However, XInputControllerOSX (Wired XBox 360) come through as [-2...2], even though they come through correctly [-1...1] from a standard **leftStick** binding.

Filtering by "up", DualShock displays [0...1] from center to top, however, XBox controller reads 0 from center to up, and [0...-2] towards down position. Filtering by "down" returns [2...0] from up to center, and remains at 0 all the way down.

Hope this is useful.

**Comment by Rene-Damm** [comment]
Ouch. That's not right for sure :)

Thanks for the details. We'll have a look.

**Comment by douglas-piconi** [comment]
fix incoming in next release.   The normalization values were messed up in the registered layout.

**Comment by douglas-piconi** [comment]
Fixed in 0.2.0-preview
